### PR TITLE
Time tracking

### PR DESCRIPTION
**Pull request by fakhirali**
Being able to track the time it takes for the system to respond.

### CODE REVIEW

File: openvoicechat/stt/base.py [hunk 1]
```diff
@@ -35,9 +40,21 @@ def listen(self) -> str:
...
+        stop_time = time.time()
+        time_diff = stop_time - start_time
+        if len(df)==0:
+            df.loc[0, 'stt'] = time_diff
+            df.at[0, 'Response'] = 'Response num: '+str(self.counter)
+        else:
+            df.loc[len(df), 'stt'] = time_diff
+            df.at[len(df), 'Response'] = 'Response num: '+str(self.counter)
+        df.to_csv(r'D:\OpenVoiceChat-master\timing.csv', index=False)
+        self.counter+=1
```
Comment (fakhirali) [inline openvoicechat/stt/base.py:57]: No need for counter. Just append a row to the dataframe. {'model':'stt', 'time taken': '1.3'}

File: openvoicechat/stt/base.py [hunk 2]
```diff
@@ -35,9 +40,21 @@ def listen(self) -> str:
         :return: transcription
         records audio using record_user and returns its transcription
         '''
-        # first transcribe
-        audio = record_user(self.silence_seconds, self.vad)
+        audio = record_user(self.silence_seconds, self.vad, self.listener)
+        df=pd.read_csv(r'D:\OpenVoiceChat-master\timing.csv')
```
Comment (fakhirali) [inline openvoicechat/stt/base.py:44]: Use relative paths

File: openvoicechat/stt/base.py [hunk 3]
```diff
@@ -35,9 +40,21 @@ def listen(self) -> str:
...
+        start_time = time.time()
         text = self.transcribe(audio)
+        stop_time = time.time()
+        time_diff = stop_time - start_time
+        if len(df)==0:
+            df.loc[0, 'stt'] = time_diff
+            df.at[0, 'Response'] = 'Response num: '+str(self.counter)
+        else:
+            df.loc[len(df), 'stt'] = time_diff
+            df.at[len(df), 'Response'] = 'Response num: '+str(self.counter)
```
Comment (fakhirali) [inline openvoicechat/stt/base.py:55]: No need to do look for this. I imagine the structure of the df to be:

| Model        | Time taken        |
|-----------------|--------------------|
| STT | 0.2 |
| LLM | 0.4 |
| TTS | 0.64 |
| STT | 0.24 |

File: openvoicechat/tts/base.py [hunk 4]
```diff
@@ -37,8 +41,8 @@ def say_text(self, text: str):
         calls run_tts and plays the audio using sounddevice.
         '''
         output = self.run_tts(text)
-        sd.play(output, samplerate=self.sample_rate)
-        sd.wait()
+        self.player.play(output, samplerate=self.sample_rate)
+        self.player.wait()
```
Comment (fakhirali) [inline openvoicechat/tts/base.py:45]: Pull from the repo's master. These should be merged from there.

File: openvoicechat/tts/base.py [hunk 5]
```diff
@@ -106,41 +110,63 @@ def say_multiple_stream(self, text_queue: queue.Queue,
         '''
         response = ''
         all_response = []
+        interrupt_text_list = []
         if audio_queue is None:
             audio_queue = queue.Queue()
         say_thread = threading.Thread(target=self.say, args=(audio_queue, listen_interruption_func))
         say_thread.start()
         while True:
+            if self.sentence_comp==0:
```
Comment (fakhirali) [inline openvoicechat/tts/base.py:119]: no need to be a class attribute. Just a first_sentence = False var would do.
Bear in mind we only care about the first sentence so no counts needed.

File: openvoicechat/tts/base.py [hunk 6]
```diff
@@ -106,41 +110,63 @@ def say_multiple_stream(self, text_queue: queue.Queue,
...
+                sentence = response
             else:
                 response += text
-                # yahan say end llm
                 if bool(re.search(self.sentence_stop_pattern, response)):
                     sentences = re.split(self.sentence_stop_pattern, response, maxsplit=1)
                     sentence = sentences[0]
                     response = sentences[1]
+                    if(self.sentence_comp==0):
+                        llm_end=time.time()
```
Comment (fakhirali) [inline openvoicechat/tts/base.py:131]: Use time.monotonic()

**Comment by haiderrzaidi** [comment]
i cloned the forked repo and tried did the changes but i am getting the following issue the first one as you see is simple i removed that chunk of arguments but then i got the second error
Screenshot from 2024-06-06 10-10-40

**Comment by fakhirali** [comment]
This should not be the case.
The record_user function does take 3 params
see [here](https://github.com/fakhirali/OpenVoiceChat/blob/df4aacb7358536a63a24077b11138c213667db86/openvoicechat/stt/utils.py#L68)
It seems like the version you have is old
the same function is [here](https://github.com/haiderrzaidi/OpenVoiceChat/blob/227248a5ea962a3c5a56cd367942428efe2e0851/openvoicechat/stt/utils.py#L60)
Let me merge them inside your fork then you can pull them from there.

**Pull request merged**